ops/framework.py

```diff
@@ -410,14 +461,18 @@ class ObjectEvents(Object):
     """Convenience type to allow defining .on attributes at class level."""
 
     handle_kind = "on"
+    if TYPE_CHECKING:
+        # to help the type checker and IDEs:
+        @property
+        def _cache(self) -> weakref.WeakKeyDictionary[Object, 'ObjectEvents']: ...  # noqa
 
-    def __init__(self, parent=None, key=None):
+    def __init__(self, parent: Optional[Object] = None, key: Optional[str] = None):
         if parent is not None:
             super().__init__(parent, key)
         else:
-            self._cache = weakref.WeakKeyDictionary()
+            self._cache = weakref.WeakKeyDictionary()  # noqa
```

Comment:
Can we just pull the `self._cache = ...` out of the else clause and add a `# type: ...` annotation and then kill the `if TYPE_CHECKING` here?  I don't see how always initializing the `_cache` hurts anything.

Reply:
Agree this is not directly related to the typing changes we are trying to make here.  I just noticed that the conditional attribute creation was a bit odd - and it really seems like always setting `self._cache` (outside the if/else) would be natural - and solves the typing awkwardness as well.  I poked around the code a bit and I didn't see any reason it needed to be conditionally defined like this.  Do you know of any code that depends on cache existing vs not as an attribute?

Reply:
No and it also feels like an odd design choice to me. Fine with refactoring this one out to simplify types.